node/block_exchange.go

```diff
@@ -0,0 +1,233 @@
+	ps := bExService.p2p.PubSub()
+	bExService.sub = goheaderp2p.NewSubscriber[*types.Block](ps, pubsub.DefaultMsgIdFn, bExService.genesis.ChainID+"-block")
+	if err := bExService.sub.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting subscriber: %w", err)
+	}
+	if _, err := bExService.sub.Subscribe(); err != nil {
+		return fmt.Errorf("error while subscribing: %w", err)
+	}
+
+	if err := bExService.blockStore.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting block store: %w", err)
+	}
+
+	_, _, network := bExService.p2p.Info()
+	var err error
+	if bExService.p2pServer, err = newBlockP2PServer(bExService.p2p.Host(), bExService.blockStore, network+"-block"); err != nil {
+		return fmt.Errorf("error while creating p2p server: %w", err)
+	}
+	if err := bExService.p2pServer.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting p2p server: %w", err)
+	}
+
+	peerIDs := bExService.p2p.PeerIDs()
+	if bExService.ex, err = newBlockP2PExchange(bExService.p2p.Host(), peerIDs, network+"-block", bExService.genesis.ChainID+"-block", bExService.p2p.ConnectionGater()); err != nil {
+		return fmt.Errorf("error while creating exchange: %w", err)
+	}
+	if err := bExService.ex.Start(bExService.ctx); err != nil {
+		return fmt.Errorf("error while starting exchange: %w", err)
+	}
+
+	if bExService.syncer, err = newBlockSyncer(bExService.ex, bExService.blockStore, bExService.sub, goheadersync.WithBlockTime(bExService.conf.BlockTime)); err != nil {
+		return fmt.Errorf("error while creating syncer: %w", err)
+	}
+
+	if bExService.isInitialized() {
+		if err := bExService.syncer.Start(bExService.ctx); err != nil {
+			return fmt.Errorf("error while starting the syncer: %w", err)
+		}
+		bExService.syncerStatus.setStarted()
+		return nil // can return here since syncer is initialized
+	}
+
+	// Look to see if trusted hash is passed, if not get the genesis block
+	var trustedBlock *types.Block
+	// Try fetching the trusted block from peers if exists
+	if len(peerIDs) > 0 {
+		if bExService.conf.TrustedHash != "" {
+			trustedHashBytes, err := hex.DecodeString(bExService.conf.TrustedHash)
+			if err != nil {
+				return fmt.Errorf("failed to parse the trusted hash for initializing the blockstore: %w", err)
+			}
+
+			if trustedBlock, err = bExService.ex.Get(bExService.ctx, header.Hash(trustedHashBytes)); err != nil {
+				return fmt.Errorf("failed to fetch the trusted block for initializing the blockStore: %w", err)
+			}
+		} else {
+			// Try fetching the genesis block if available, otherwise fallback to blocks
+			if trustedBlock, err = bExService.ex.GetByHeight(bExService.ctx, uint64(bExService.genesis.InitialHeight)); err != nil {
+				// Full/light nodes have to wait for aggregator to publish the genesis block
+				// proposing aggregator can init the store and start the syncer when the first block is published
+				bExService.logger.Info("failed to fetch the genesis block", "error", err)
```

Comment:
lets return error here, which will prevent the start of secondary aggregator and full nodes.